File: SQL/Select/SQLSelectGroupByFields.vb
' ___________________________________________________
'
'  (c) Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLSelectGroupByFields
        Implements Collections.Generic.IEnumerable(Of SQLSelectGroupByField)

        Private pobjGroupByFields As New Collections.Generic.List(Of SQLSelectGroupByField)

        Public Sub New()

            MyBase.New()

        End Sub

        Public Function Add() As SQLSelectGroupByField

            Return Add("", Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String) As SQLSelectGroupByField

            Return Add(strFieldName, Nothing)

        End Function

        Public Function Add( _
            ByVal strFieldName As String, _
            ByVal objTable As SQLSelectTable) As SQLSelectGroupByField

            Dim objFieldOrder As New SQLSelectGroupByField(New SQLFieldExpression(objTable, strFieldName))

            pobjGroupByFields.Add(objFieldOrder)

            Return objFieldOrder

        End Function

        Public Function Add( _
            ByVal objExpression As SQLExpression) As SQLSelectGroupByField

            Dim objFieldOrder As New SQLSelectGroupByField(objExpression)

            pobjGroupByFields.Add(objFieldOrder)

            Return objFieldOrder

        End Function

        Default Public ReadOnly Property Item(ByVal strFieldName As String) As SQLSelectGroupByField
            Get

                Return Me(FieldNameIndex(strFieldName))

            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal intIndex As Integer) As SQLSelectGroupByField
            Get

                Return pobjGroupByFields.Item(intIndex)

            End Get
        End Property

        Public ReadOnly Property Count() As Integer
            Get

                Return pobjGroupByFields.Count()

            End Get
        End Property

        Friend ReadOnly Property SQL(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = String.Empty

                For intIndex As Integer = 0 To Me.Count - 1
                    strSQL &= Me.Item(intIndex).SQL(eConnectionType)
                    If intIndex <> Me.Count - 1 Then
                        strSQL &= ", "
                    End If
                Next

                Return strSQL

            End Get
        End Property

        Public Function Exists(ByVal strFieldName As String) As Boolean

            Return FieldNameIndex(strFieldName) >= 0

        End Function

        Public Sub Delete(ByRef objGroupByField As SQLSelectGroupByField)

            If Not pobjGroupByFields.Contains(objGroupByField) Then
                Throw New IndexOutOfRangeException
            End If

            pobjGroupByFields.Remove(objGroupByField)
            objGroupByField = Nothing

        End Sub

        Private Function FieldNameIndex(ByVal strFieldName As String) As Integer

            Dim objGroupByField As SQLSelectGroupByField

            strFieldName = strFieldName.Trim()

            For intIndex As Integer = 0 To Me.Count - 1
                objGroupByField = DirectCast(pobjGroupByFields.Item(intIndex), SQLSelectGroupByField)
                If TypeOf objGroupByField.Expression Is SQLFieldExpression Then
                    If String.Compare(strFieldName, DirectCast(objGroupByField.Expression, SQLFieldExpression).Name, ignoreCase:=True) = 0 Then
                        Return intIndex
                    End If
                End If
            Next

            Return -1

        End Function

        Public ReadOnly Property IsEmpty As Boolean
            Get

                Return pobjGroupByFields.Count = 0

            End Get
        End Property

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator

            Return pobjGroupByFields.GetEnumerator

        End Function

        Private Function GetEnumerator1() As System.Collections.Generic.IEnumerator(Of SQLSelectGroupByField) Implements System.Collections.Generic.IEnumerable(Of SQLSelectGroupByField).GetEnumerator

            Return pobjGroupByFields.GetEnumerator

        End Function

    End Class

End Namespace

File: SQL/Select/SQLSelectGroupByFields.cs
// _________________________________________________________________________
//
//  (c) Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Collections.Generic;

namespace DatabaseObjects.SQL
{
    public class SQLSelectGroupByFields : IEnumerable<SQLSelectGroupByField>
    {
        private List<SQLSelectGroupByField> pobjGroupByFields = new List<SQLSelectGroupByField>();

        public SQLSelectGroupByFields()
        {
        }

        public SQLSelectGroupByField Add()
        {
            return Add("", null);
        }

        public SQLSelectGroupByField Add(string strFieldName)
        {
            return Add(strFieldName, null);
        }

        public SQLSelectGroupByField Add(string strFieldName, SQLSelectTable objTable)
        {
            SQLSelectGroupByField objFieldOrder = new SQLSelectGroupByField(new SQLFieldExpression(objTable, strFieldName));

            pobjGroupByFields.Add(objFieldOrder);

            return objFieldOrder;
        }

        public SQLSelectGroupByField Add(SQLExpression objExpression)
        {
            SQLSelectGroupByField objFieldOrder = new SQLSelectGroupByField(objExpression);

            pobjGroupByFields.Add(objFieldOrder);

            return objFieldOrder;
        }

        public SQLSelectGroupByField this[string strFieldName]
        {
            get
            {
                return this[FieldNameIndex(strFieldName)];
            }
        }

        public SQLSelectGroupByField this[int intIndex]
        {
            get
            {
                return pobjGroupByFields[intIndex];
            }
        }

        public int Count
        {
            get
            {
                return pobjGroupByFields.Count;
            }
        }
		
        public bool Exists(string strFieldName)
        {
            return FieldNameIndex(strFieldName) >= 0;
        }

        public void Delete(ref SQLSelectGroupByField objGroupByField)
		{
			if (!pobjGroupByFields.Contains(objGroupByField))
				throw new IndexOutOfRangeException();
				
			pobjGroupByFields.Remove(objGroupByField);
			objGroupByField = null;
		}

        private int FieldNameIndex(string strFieldName)
        {
            SQLSelectGroupByField objGroupByField;

            for (int intIndex = 0; intIndex < this.Count; intIndex++)
            {
                objGroupByField = (SQLSelectGroupByField)(pobjGroupByFields[intIndex]);
                if (objGroupByField.Expression is SQLFieldExpression)
                {
                    if (string.Compare(strFieldName, ((SQLFieldExpression)objGroupByField.Expression).Name, true) == 0)
                        return intIndex;
                }
            }

            return -1;
        }

        public bool IsEmpty
        {
            get
            {
                return pobjGroupByFields.Count == 0;
            }
        }

        public System.Collections.IEnumerator GetEnumerator()
        {
            return pobjGroupByFields.GetEnumerator();
        }

        IEnumerator<SQLSelectGroupByField> IEnumerable<SQLSelectGroupByField>.GetEnumerator()
        {
            return pobjGroupByFields.GetEnumerator();
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return pobjGroupByFields.GetEnumerator();
        }
    }
}